### PR TITLE
Absolute attrPath notation ("flakeref#.attrPath")

### PR DESCRIPTION
**Pull request by tomberek**
# Motivation
Allow referring to a flake attribute path without accidental default prefix'ing.

Introduce a syntax of "flakeref#.attrPath" where the "." indicated the root of the flake outputs. A natural consequence of this is no defaults and no default prefix insertions (eg: "legacyPackages.SYSTEM")

# Context
Closes: https://github.com/NixOS/nix/issues/8665
Continuation: https://github.com/NixOS/nix/pull/8715

# Checklist for maintainers

- [ ] Might be better to represent this as a variant of Installable?
- [ ] Would the attrPaths in profile manifest.json's benefit from this?

Maintainers: tick if completed or explain if not relevant

 - [x] agreed on idea
 - [ ] agreed on implementation strategy
 - [x] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [x] documentation in the manual
 - [ ] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [x] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

File: src/libcmd/installable-flake.cc [hunk 1]
```diff
@@ -28,6 +28,11 @@ namespace nix {
 std::vector<std::string> InstallableFlake::getActualAttrPaths()
 {
     std::vector<std::string> res;
+    if (attrPaths.size() == 1 && attrPaths.front().starts_with(".")){
```
Comment (roberth) [inline src/libcmd/installable-flake.cc:31]: I've read the method docs, but I still don't really understand what this method does - crucially why it treats the first element specially.
We're going to need some documentation and/or refactoring here.

Comment (tomberek) [reply src/libcmd/installable-flake.cc:31]: Agreed, this is probably incorrect. Or at least needs justification.

**Comment by iFreilicht** [comment]
Would be great if you could add some tests to verify this new behavior.

File: src/libcmd/installable-flake.cc (hunk 1 above)
Comment (tomberek) [reply src/libcmd/installable-flake.cc:31]: note: make this an assert

Comment (tomberek) [reply src/libcmd/installable-flake.cc:31]: This function seems to be called at times with the defaultPaths already present, namely when there is no attrPath component to the installable. (due to: https://github.com/NixOS/nix/blob/master/src/libcmd/installable-flake.cc#L76)

A possibility is to move the logic into the constructor. Something like:
```
    : InstallableValue(state),
      flakeRef(flakeRef),
      attrPaths(fragment == "" ? attrPaths : Strings{(std::string) fragment}),
      prefixes(fragment == "" || !fragment.starts_with(".") ? Strings{} : prefixes),
      extendedOutputsSpec(std::move(extendedOutputsSpec)),
      lockFlags(lockFlags)
```

Comment (roberth) [reply src/libcmd/installable-flake.cc:31]: "An installable flake has attribute paths and prefixes" is a completely nonsensical statement to me, so I have no idea what the implications of this change would be without digging through everything.

Resolving installables is a procedure, not data, so I don't think attribute paths and prefixes should be in the same object in the first place.

So I guess this object is a representation of the internal state of that procedure, which is "fine", and what you're suggestion might work, And I guess what I'm trying to say is that a more functional or procedural style would help for cleaning up the existing code.

**Review by roberth — APPROVED** [review]

**Comment by roberth** [comment]
@tomberek ready for merge?

**Comment by tomberek** [comment]
@roberth i think so. This part of the codebase does need a cleanup and refactor, but I don’t think that should block this. I do appreciate an  opinion/review - to see if I’ve missed something, an interaction or complexity I missed, consider an alternative approach. 

Anything I should change/add/remove?

**Pull request merged**